src/core/meshcop/tcat_agent.cpp

```diff
@@ -0,0 +1,515 @@
+ *     names of its contributors may be used to endorse or promote products
+ *     derived from this software without specific prior written permission.
+ *
+ *  THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+ *  AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ *  IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+ *  ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
+ *  LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+ *  CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ *  SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ *  INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ *  CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ *  ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ *  POSSIBILITY OF SUCH DAMAGE.
+ */
+
+/**
+ * @file
+ *   This file implements the TCAT Agent service.
+ */
+
+#include "tcat_agent.hpp"
+
+#if OPENTHREAD_CONFIG_BLE_TCAT_ENABLE
+
+#include <stdio.h>
+
+#include "common/array.hpp"
+#include "common/code_utils.hpp"
+#include "common/debug.hpp"
+#include "common/encoding.hpp"
+#include "common/instance.hpp"
+#include "common/locator_getters.hpp"
+#include "common/string.hpp"
+#include "radio/radio.hpp"
+#include "thread/thread_netif.hpp"
+#include "thread/uri_paths.hpp"
+#include "utils/otns.hpp"
+
+namespace ot {
+namespace MeshCoP {
+
+RegisterLogModule("TcatAgent");
+
+bool TcatAgent::VendorInfo::IsValid(void) const
+{
+    return mProvisioningUrl == nullptr || IsValidUtf8String(mProvisioningUrl) || mPskdString != nullptr;
+}
+
+TcatAgent::TcatAgent(Instance &aInstance)
+    : InstanceLocator(aInstance)
+    , mVendorInfo(nullptr)
+    , mCurrentApplicationLayer(kMessageTypeNone)
+    , mState(kStateDisabled)
+    , mAlreadyCommissioned(false)
+    , mCommissionerHasNetworkName(false)
+    , mCommissionerHasDomainName(false)
+    , mCommissionerHasExtendedPanId(false)
+{
+    mJoinerPskd.Clear();
+    mCurrentServiceName[0] = 0;
+}
+
+Error TcatAgent::Start(const TcatAgent::VendorInfo &aVendorInfo,
+                       AppDataReceiveCallback       aAppDataReceiveCallback,
+                       JoinCallback                 aHandler,
+                       void                        *aContext)
+{
+    Error error = kErrorNone;
+
+    LogInfo("Starting");
+
+    VerifyOrExit(aVendorInfo.IsValid(), error = kErrorInvalidArgs);
+    SuccessOrExit(error = mJoinerPskd.SetFrom(aVendorInfo.mPskdString));
+
+    mAppDataReceiveCallback.Set(aAppDataReceiveCallback, aContext);
+    mJoinCallback.Set(aHandler, aContext);
+
+    mVendorInfo              = &aVendorInfo;
+    mCurrentApplicationLayer = kMessageTypeNone;
+    mState                   = kStateEnabled;
+    mAlreadyCommissioned     = false;
+
+exit:
+    LogError("start TCAT agent", error);
```

Comment:
Suggest defining `LogError()` as a method in `TcatAgent` directly? 
(maybe similar to this PR -> https://github.com/openthread/openthread/pull/9432)

This is currently using the common `MeshCop::LogError()` function. The problem with using this function is it wont use the proper "log module" (which is defined as `RegisterLogModule("TcatAgent") here)` which makes it harder during debugging to filer out logs for a specific module.

We may remove the common `LogError()` soon in favor of adding it to each class directly so the correct log module is used in future.

Reply:
Done